# Notebook 1 - Read and join the tables

Builds one table with a single row per order out of the Olist database on `localhost:5433`
and writes it to `artifacts/01_ml_table.parquet`.

`order_items` and `order_payments` hold several rows per order, so they are aggregated before
they are joined.

## 0. Setup

In [1]:
import pandas as pd
from sqlalchemy import create_engine, text
from pathlib import Path

DB_URL = "postgresql+psycopg://olist:olist123@localhost:5433/olist"
engine = create_engine(DB_URL)

ARTIFACTS = Path("..") / "artifacts"
ARTIFACTS.mkdir(exist_ok=True)

def q(sql: str) -> pd.DataFrame:
    '''Run a SQL query and return a DataFrame. Shorthand used throughout this notebook.'''
    return pd.read_sql(text(sql), engine)
#TEST
q("SELECT version() AS postgres_version")

,postgres_version
0,PostgreSQL 16.14 (Debian 16.14-1.pgdg13+1) on ...


## 1. Table inventory

Row counts per table, and whether the natural key is really unique.

In [2]:
tables = ["orders", "customers", "order_items", "order_payments",
          "order_reviews", "products", "sellers", "geolocation",
          "product_category_name_translation"]

inventory = pd.DataFrame(
    [(t, q(f"SELECT count(*) AS n FROM {t}")["n"].iloc[0]) for t in tables],
    columns=["table", "rows"]
).sort_values("rows", ascending=False, ignore_index=True)

print(f"Total rows: {inventory['rows'].sum():,}")
inventory

Total rows: 1,550,922


,table,rows
0,geolocation,1000163
1,order_items,112650
2,order_payments,103886
3,orders,99441
4,customers,99441
5,order_reviews,99224
6,products,32951
7,sellers,3095
8,product_category_name_translation,71


### Row grain

Where `count(*)` differs from the number of distinct `order_id`, the table holds several rows per
order and joining it directly would multiply our rows.

In [3]:
grain = q('''
SELECT 'orders'         AS tbl, count(*) AS rows, count(DISTINCT order_id) AS distinct_orders FROM orders
UNION ALL SELECT 'order_items',    count(*), count(DISTINCT order_id) FROM order_items
UNION ALL SELECT 'order_payments', count(*), count(DISTINCT order_id) FROM order_payments
UNION ALL SELECT 'order_reviews',  count(*), count(DISTINCT order_id) FROM order_reviews
''')

grain["rows_per_order"] = (grain["rows"] / grain["distinct_orders"]).round(3)
grain["needs_aggregation"] = grain["rows"] != grain["distinct_orders"]
grain

,tbl,rows,distinct_orders,rows_per_order,needs_aggregation
0,order_items,112650,98666,1.142,True
1,orders,99441,99441,1.000,False
2,order_payments,103886,99440,1.045,True
3,order_reviews,99224,98673,1.006,True


## 2. Coverage

Is every order present in the child tables? If not, an `INNER JOIN` drops the missing ones with no
error and no warning.

In [4]:
coverage = q('''
SELECT (SELECT count(*) FROM orders)                        AS orders_total,
       (SELECT count(DISTINCT order_id) FROM order_items)   AS present_in_items,
       (SELECT count(DISTINCT order_id) FROM order_payments) AS present_in_payments,
       (SELECT count(DISTINCT order_id) FROM order_reviews)  AS present_in_reviews
''').T.rename(columns={0: "count"})

coverage["missing_orders"] = coverage.loc["orders_total", "count"] - coverage["count"]
coverage

,count,missing_orders
orders_total,99441,0
present_in_items,98666,775
present_in_payments,99440,1
present_in_reviews,98673,768


775 orders have no line in `order_items`. Every join below is therefore a `LEFT JOIN` from
`orders`, and the row count has to stay at 99,441 - checked in section 8.

## 3. Aggregating the order lines

`shipping_limit_date` is the seller's deadline to hand the parcel to the carrier. It is set when
the order is created, so it is available at prediction time - unlike `order_delivered_carrier_date`,
which records the actual handover and is leakage.

In [5]:
items_agg = q('''
SELECT order_id,
       count(*)                    AS n_items,          -- how many lines in the order
       count(DISTINCT product_id)  AS n_products,       -- how many distinct products
       count(DISTINCT seller_id)   AS n_sellers,        -- how many distinct sellers
       sum(price)                  AS items_price_total,
       max(price)                  AS items_price_max,
       sum(freight_value)          AS freight_total,
       max(freight_value)          AS freight_max,
       min(shipping_limit_date)    AS shipping_limit_first
FROM order_items
GROUP BY order_id
''')

print(f"{len(items_agg):,} orders after aggregation")
items_agg.head()

98,666 orders after aggregation


,order_id,n_items,n_products,n_sellers,items_price_total,items_price_max,freight_total,freight_max,shipping_limit_first
0,00010242fe8c5a6d1ba2dd792cb16214,1,1,1,58.90,58.90,13.29,13.29,2017-09-19 09:45:35
1,00018f77f2f0320c557190d7a144bdd3,1,1,1,239.90,239.90,19.93,19.93,2017-05-03 11:05:13
2,000229ec398224ef6ca0657da4fc703e,1,1,1,199.00,199.00,17.87,17.87,2018-01-18 14:48:30
3,00024acbcdf0a6daa1e931b038114c75,1,1,1,12.99,12.99,12.79,12.79,2018-08-15 10:10:18
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,1,1,199.90,199.90,18.14,18.14,2017-02-13 13:57:51


### Orders with several sellers

Not every order has a single seller, which makes "the seller's state" ambiguous.

In [6]:
q('''
SELECT n_sellers, count(*) AS orders
FROM (SELECT order_id, count(DISTINCT seller_id) AS n_sellers FROM order_items GROUP BY order_id) t
GROUP BY n_sellers ORDER BY n_sellers
''')

,n_sellers,orders
0,1,97388
1,2,1219
2,3,54
3,4,3
4,5,2


The highest-priced line is taken as the main line, and the seller, category, state and zip prefix
are read from it. `order_item_id` breaks price ties so the result is deterministic. `n_sellers` is
kept as a column of its own.

`DISTINCT ON` returns the first row per `order_id` under the given `ORDER BY`.

In [7]:
main_line = q('''
SELECT DISTINCT ON (oi.order_id)
       oi.order_id,
       oi.seller_id              AS main_seller_id,
       s.seller_state            AS main_seller_state,
       s.seller_zip_code_prefix  AS main_seller_zip,
       p.product_category_name   AS main_category
FROM order_items oi
JOIN sellers  s USING (seller_id)
JOIN products p USING (product_id)
ORDER BY oi.order_id, oi.price DESC, oi.order_item_id   -- most expensive first; order_item_id breaks ties
''')

print(f"{len(main_line):,} orders")
main_line.head()

98,666 orders


,order_id,main_seller_id,main_seller_state,main_seller_zip,main_category
0,00010242fe8c5a6d1ba2dd792cb16214,48436dade18ac8b2bce089ec2a041202,SP,27277,cool_stuff
1,00018f77f2f0320c557190d7a144bdd3,dd7ddc04e1b6c2c614352b383efe2d36,SP,03471,pet_shop
2,000229ec398224ef6ca0657da4fc703e,5b51032eddd242adc84c38acab88f23d,MG,37564,moveis_decoracao
3,00024acbcdf0a6daa1e931b038114c75,9d7a1d34a5052409006425275ba1c2b4,SP,14403,perfumaria
4,00042b26cf59d7ce69dfabb4e55b4fd9,df560393f3a51e74553ab94004ba5c87,PR,87900,ferramentas_jardim


## 4. Product attributes

Weight and volume are aggregated over the whole order, since the shipment
travels together.

In [8]:
product_agg = q('''
SELECT oi.order_id,
       sum(p.product_weight_g)                          AS weight_g_total,
       max(p.product_weight_g)                          AS weight_g_max,
       sum(p.product_length_cm * p.product_height_cm
             * p.product_width_cm)                      AS volume_cm3_total,
       count(DISTINCT p.product_category_name)          AS n_categories,
       avg(p.product_photos_qty)                        AS photos_avg
FROM order_items oi
JOIN products p USING (product_id)
GROUP BY oi.order_id
''')

print(f"{len(product_agg):,} orders")
product_agg.head()

98,666 orders


,order_id,weight_g_total,weight_g_max,volume_cm3_total,n_categories,photos_avg
0,00010242fe8c5a6d1ba2dd792cb16214,650.0,650.0,3528.0,1,4.0
1,00018f77f2f0320c557190d7a144bdd3,30000.0,30000.0,60000.0,1,2.0
2,000229ec398224ef6ca0657da4fc703e,3050.0,3050.0,14157.0,1,2.0
3,00024acbcdf0a6daa1e931b038114c75,200.0,200.0,2400.0,1,1.0
4,00042b26cf59d7ce69dfabb4e55b4fd9,3750.0,3750.0,42000.0,1,1.0


## 5. Payments

`mode() WITHIN GROUP` gives the dominant payment type of the order.

In [9]:
payments_agg = q('''
SELECT order_id,
       count(*)                                          AS n_payments,
       sum(payment_value)                                AS payment_total,
       max(payment_installments)                         AS installments_max,
       mode() WITHIN GROUP (ORDER BY payment_type)       AS main_payment_type
FROM order_payments
GROUP BY order_id
''')

print(f"{len(payments_agg):,} orders")
payments_agg.head()

99,440 orders


,order_id,n_payments,payment_total,installments_max,main_payment_type
0,00010242fe8c5a6d1ba2dd792cb16214,1,72.19,2,credit_card
1,00018f77f2f0320c557190d7a144bdd3,1,259.83,3,credit_card
2,000229ec398224ef6ca0657da4fc703e,1,216.87,5,credit_card
3,00024acbcdf0a6daa1e931b038114c75,1,25.78,2,credit_card
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,218.04,3,credit_card


## 6. Geography

`geolocation` holds 1,000,163 rows for 19,015 zip prefixes, about 53 points each. It is reduced to
one point per prefix, using the median: the table contains coordinates outside
Brazil.

In [10]:
geo = q('''
SELECT geolocation_zip_code_prefix                                   AS zip_prefix,
       percentile_cont(0.5) WITHIN GROUP (ORDER BY geolocation_lat)  AS lat,
       percentile_cont(0.5) WITHIN GROUP (ORDER BY geolocation_lng)  AS lng
FROM geolocation
GROUP BY geolocation_zip_code_prefix
''')

print(f"1,000,163 rows  ->  {len(geo):,} zip prefixes")
geo.head()

1,000,163 rows  ->  19,015 zip prefixes


,zip_prefix,lat,lng
0,01001,-23.550381,-46.634027
1,01002,-23.548551,-46.635072
2,01003,-23.548977,-46.635313
3,01004,-23.549535,-46.634771
4,01005,-23.549612,-46.636532


## 7. Base table

`orders` joined to `customers`, one to one. The post-purchase timestamps are carried along because
notebook 2 needs `order_delivered_customer_date` to build the label; they are dropped before the
features in notebook 5.

In [11]:
base = q('''
SELECT o.order_id,
       o.customer_id,
       o.order_status,
       o.order_purchase_timestamp,
       o.order_estimated_delivery_date,
       -- the columns below are not available at purchase time: label only, dropped in notebook 5
       o.order_approved_at,
       o.order_delivered_carrier_date,
       o.order_delivered_customer_date,
       c.customer_unique_id,
       c.customer_zip_code_prefix,
       c.customer_city,
       c.customer_state
FROM orders o
JOIN customers c USING (customer_id)
''')

print(f"{len(base):,} orders")
base.head(3)

99,441 orders


,order_id,customer_id,order_status,order_purchase_timestamp,order_estimated_delivery_date,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-18,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,7c396fd4830fd04220f754e42b4e5bff,03149,sao paulo,SP
1,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-09-04,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,3a653a41f6f9fc3d2a113cf8398680e8,75265,vianopolis,GO
2,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-12-15,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,7c142cf63193a1473d2e66489a9ae977,59296,sao goncalo do amarante,RN


## 8. Assembling the ML table

Every join is `how="left"` from `base`, with `validate="one_to_one"` so that a non-unique key raises
instead of silently duplicating rows. The row count is printed after each step.

In [12]:
ml = base.copy()

for name, part in [("items", items_agg), ("main_line", main_line),
                   ("products", product_agg), ("payments", payments_agg)]:
    before = len(ml)
    ml = ml.merge(part, on="order_id", how="left", validate="one_to_one")
    print(f"+ {name:<10} -> {len(ml):,} rows  (was {before:,})")

+ items      -> 99,441 rows  (was 99,441)


+ main_line  -> 99,441 rows  (was 99,441)


+ products   -> 99,441 rows  (was 99,441)
+ payments   -> 99,441 rows  (was 99,441)


In [13]:
# customer coordinates and main-seller coordinates, both taken from the aggregated geo table
ml = (ml
      .merge(geo.rename(columns={"zip_prefix": "customer_zip_code_prefix",
                                 "lat": "customer_lat", "lng": "customer_lng"}),
             on="customer_zip_code_prefix", how="left", validate="many_to_one")
      .merge(geo.rename(columns={"zip_prefix": "main_seller_zip",
                                 "lat": "seller_lat", "lng": "seller_lng"}),
             on="main_seller_zip", how="left", validate="many_to_one"))

print(f"After adding coordinates: {len(ml):,} rows x {ml.shape[1]} columns")

After adding coordinates: 99,441 rows x 37 columns


## 9. Checks

In [14]:
n_orders = q("SELECT count(*) AS n FROM orders")["n"].iloc[0]

assert len(ml) == n_orders,          f"row count {len(ml)} != order count {n_orders}"
assert ml["order_id"].is_unique,     "duplicate order_id - an aggregation failed somewhere"
assert ml["order_id"].notna().all(), "empty order_id found"

print(f"OK  {len(ml):,} rows = {n_orders:,} orders")
print(f"OK  order_id is unique")
print(f"OK  {ml.shape[1]} columns")

OK  99,441 rows = 99,441 orders
OK  order_id is unique
OK  37 columns


In [15]:
# missing values: expected in the columns of orders that have no children - we want to see them, not hide them
missing = (ml.isna().sum().to_frame("missing")
             .assign(pct=lambda d: (100 * d["missing"] / len(ml)).round(2))
             .query("missing > 0")
             .sort_values("missing", ascending=False))
missing

,missing,pct
order_delivered_customer_date,2965,2.98
main_category,2190,2.20
photos_avg,2164,2.18
order_delivered_carrier_date,1783,1.79
seller_lat,991,1.00
seller_lng,991,1.00
weight_g_total,791,0.80
volume_cm3_total,791,0.80
weight_g_max,791,0.80
freight_total,775,0.78


## 10. Saving

Parquet keeps the column dtypes, so notebook 2 reads the timestamps back as
timestamps.

In [16]:
out = ARTIFACTS / "01_ml_table.parquet"
ml.to_parquet(out, index=False)

size_mb = out.stat().st_size / 1024**2
print(f"Saved: {out}  ({size_mb:.1f} MB)")
print(f"       {len(ml):,} rows x {ml.shape[1]} columns")

Saved: ..\artifacts\01_ml_table.parquet  (17.3 MB)
       99,441 rows x 37 columns


---

**Result:** `artifacts/01_ml_table.parquet`, 99,441 rows, one per order.

Decisions taken here:

- `LEFT JOIN` from `orders` throughout - 775 orders have no lines at all.
- Main line = highest-priced line; `n_sellers` kept separately.
- Median coordinates per zip prefix, not the mean.
- Post-purchase timestamps carried for the label, dropped in notebook 5.
- Distance is not computed here; it belongs with the features.